src/query/sortResults.ts

```diff
@@ -0,0 +1,91 @@
+import debug from 'debug'
+import get from 'lodash.get'
+import { InternalEntity } from 'src/glossary'
+import { OrderBy, SortDirection } from './queryTypes'
+
+const log = debug('sortResults')
+
+type FlatSortCriteria = [string[], SortDirection]
+
+export function warnOnIneffectiveSortingKeys(obj: Record<string, any>): void {
+  const [mainCriteria, ...siblings] = Object.keys(obj)
+
+  if (siblings.length > 0) {
+    console.warn(
+      'Sorting by "%s" has no effect: already sorted by "%s".',
+      siblings.join(','),
+      mainCriteria,
+    )
+  }
+}
+
+function flattenSortCriteria(
+  orderBy: OrderBy<any>[],
+  propertyPath: string[] = [],
+): FlatSortCriteria[] {
+  log('flattenSortCriteria:', orderBy, propertyPath)
+
+  return orderBy.reduce<FlatSortCriteria[]>((criteria, properties) => {
+    warnOnIneffectiveSortingKeys(properties)
+
+    // Multiple properties in a single criteria object are forbidden.
+    // Use the list of criteria objects for multi-criteria sort.
+    const property = Object.keys(properties)[0]
+    const sortDirection = properties[property]!
+    const path = propertyPath.concat(property)
+    log({ property, sortDirection, path })
+
+    // Recursively flatten order criteria when referencing
+    // relational properties.
+    const newCriteria =
+      typeof sortDirection === 'object'
+        ? flattenSortCriteria([sortDirection], path)
+        : [[path, sortDirection]]
+
+    log('pushing new criteria:', newCriteria)
+    return criteria.concat(newCriteria as any)
```

Comment:
This 'as' is pretty gnarly

If you change line 43 to:
```ts
: [[path, sortDirection]] as FlatSortCriteria[]
```

Then you can remove this `as any` here